FixMath.h

```diff
@@ -0,0 +1,1287 @@
+  }
+
+  /** Safe and optimal left shift. The returned type will be adjusted accordingly
+      @param op The shift number
+      @return The result of the shift as a UFixMath of bigger size.
+  */
+  template<int8_t op>
+  UFixMath<NI+op,MozziPrivate::mozziMax(NF-op,0),MozziPrivate::RANGESHIFT(NF,op,RANGE)> sL()
+  {
+    return UFixMath<NI+op,MozziPrivate::mozziMax(NF-op,0)>(internal_value,true);
+  }
+
+  
+  //////// COMPARISON OVERLOADS
+  /** Comparison with another UFixMath.
+      @param op A UFixMath
+      @return true if this is bigger than op, false otherwise
+  */
+  template<int8_t _NI, int8_t _NF>
+  bool operator> (const UFixMath<_NI,_NF>& op) const
+  {
+    constexpr int8_t new_NI = MozziPrivate::mozziMax(NI, _NI);
+    constexpr int8_t new_NF = MozziPrivate::mozziMax(NF, _NF);
+    UFixMath<new_NI,new_NF> left(*this);
+    UFixMath<new_NI,new_NF> right(op);
+    return left.asRaw()>right.asRaw();
+  }
+
+  /** Comparison with another UFixMath.
+      @param op A UFixMath
+      @return true if this is smaller than op, false otherwise
+  */
+  template<int8_t _NI, int8_t _NF>
+  bool operator< (const UFixMath<_NI,_NF>& op) const
+  {
+    return op > *this;
+  }
+
+
+  /** Comparison with another UFixMath.
+      @param op A UFixMath
+      @return true if this equal to op, false otherwise
+  */
+  template<int8_t _NI, int8_t _NF>
+  bool operator== (const UFixMath<_NI,_NF>& op) const
+  {
+    constexpr int8_t new_NI = MozziPrivate::mozziMax(NI, _NI);
+    constexpr int8_t new_NF = MozziPrivate::mozziMax(NF, _NF);
+    UFixMath<new_NI,new_NF> left(*this);
+    UFixMath<new_NI,new_NF> right(op);
+    return left.asRaw()==right.asRaw();
+  }
+
+  /** Comparison with another UFixMath.
+      @param op A UFixMath
+      @return true if this not equal to op, false otherwise
+  */
+  template<int8_t _NI, int8_t _NF>
+  bool operator!= (const UFixMath<_NI,_NF>& op) const
+  {
+    constexpr int8_t new_NI = MozziPrivate::mozziMax(NI, _NI);
+    constexpr int8_t new_NF = MozziPrivate::mozziMax(NF, _NF);
+    UFixMath<new_NI,new_NF> left(*this);
+    UFixMath<new_NI,new_NF> right(op);
+    return left.asRaw()!=right.asRaw();
+  }
+  
+  /** Returns the number as a SFixMath of same range and precision. This is more optimized than a cast.
+      @return a SFixMath 
+  */
+  SFixMath<NI,NF,RANGE> asSFix()
+  {
+    return SFixMath<NI,NF,RANGE>(internal_value,true);
+  }
+
+  /** Returns the value as floating point number.
+      @return The floating point value.
+  */
+  float asFloat() { return (static_cast<float>(internal_value)) / (next_greater_type(1)<<NF); }
+  
+  /** Returns the internal integer value
+      @return the internal value
+  */
+  internal_type asRaw() const { return internal_value; }
+  
+  /** The number of bits used to encode the integral part.
+      @return The number of bits used to encode the integral part.
+  */
+  int8_t getNI(){return NI;}
```

Comment:
Should be declared const, too (and constexpr, but let's delay worrying about constexpr to after this is merged).

Reply:
Good catch, and a few others too!